api/src/main/java/org/apache/iceberg/catalog/SupportsCatalogTransactions.java

```diff
@@ -0,0 +1,137 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.catalog;
+
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * Catalog methods for working with catalog-level transactions.
+ *
+ * <p>Catalog implementations are not required to support catalog-level transactional state.
+ * If they do, they may support one or more {@code IsolationLevel}s and one or more
+ * {@code LockingMode}s.
+ */
+public interface SupportsCatalogTransactions {
+
+  /**
+   * The level of isolation for a transaction.
+   */
+  enum IsolationLevel {
+
+    /**
+     * Only read committed data. Reading the same table multiple times
+     * may result in different committed reads for each read.
+     */
+    READ_COMMITTED,
+
+    /**
+     * Only read committed data. Reading the same table multiple times will
+     * result in the same view of those tables.
+     */
+    REPEATED_READ,
+
+    /**
+     * Only read committed data. A commit will only succeed if there have
+     * been no changes to data read during the course of the transaction
+     * prior to the commit operation.
+     */
+    SERIALIZABLE;
+  }
+
+  /**
+   * The type of locking mode used by the transaction.
+   */
+  enum LockingMode {
+
+    /**
+     * Pessimistically lock tables. In this situation, each table being worked on will grab a lock.
+     * This mode will typically result in a better chance for transactions to complete at the cost
+     * of throughput and resource consumption.
+     */
+    PESSIMISTIC,
+
+    /**
+     * Run the transaction with an optimistic behavior. This assumes that most operations will not
+     * conflict. It typically increases concurrency and throughput while reducing resource
+     * consumption.
+     */
+    OPTIMISTIC;
+  }
+
+  /**
+   * Get the list of {@link LockingMode}s this Catalog supports.
+   * @return A set of locking modes supported.
+   */
+  default Set<LockingMode> lockingModes() {
+    return Collections.emptySet();
+  }
+
+  /**
+   * Get the list of {@link IsolationLevel}s this Catalog supports.
+   * @return A set of locking modes supported.
+   */
+  default Set<IsolationLevel> isolationLevels() {
+    return Collections.emptySet();
+  }
+
+  /**
+   * Start a new transaction and return a {@link TransactionalCatalog} that supports
+   * cross table operations. If a {@link Catalog} supports {@link OPTIMISTIC}
+   * locking mode, this will use that mode. If it only supports {@link PESSIMISTIC},
+   * it will use that mode with default values for lock times.
+   *
+   * @param isolationLevel The {@link IsolationLevel} to use for the transaction.
+   * @return A Catalog with an open transaction.
+   * @throws IllegalArgumentException If the IsolationLevel is not supported.
+   */
+  TransactionalCatalog createTransaction(IsolationLevel isolationLevel);
```

Comment:
Another reason to automatically lock tables is that this would work without supplying tables here. If we require tables to be passed in to lock them, then this would need a list of tables. Not supplying them would mean that this can't fall back to pessimistic.

Reply:
Per my comments above. The table list is optional if you want to optimize a pessimistic operation. Not a pre-commitment that those are the only tables you'll access.